lib/Core/AddressSpace.cpp

```diff
@@ -336,12 +336,26 @@ bool AddressSpace::copyInConcretes() {
 bool AddressSpace::copyInConcrete(const MemoryObject *mo, const ObjectState *os,
                                   uint64_t src_address) {
   auto address = reinterpret_cast<std::uint8_t*>(src_address);
-  if (memcmp(address, os->concreteStore, mo->size) != 0) {
+  // Check if the underlying representation has been changed externally
+  if (std::memcmp(address, os->concreteStore, mo->size) != 0) {
     if (os->readOnly) {
       return false;
     } else {
       ObjectState *wos = getWriteable(mo, os);
-      memcpy(wos->concreteStore, address, mo->size);
+      if (wos->unflushedMask) {
+        // Handle the case, that objects have been modified by external
+        // functions. Do a byte-by-byte comparison and update if needed.
+        for (size_t i = 0, ie = mo->size; i < ie; ++i) {
+          u_int8_t updated_value = *(address + i);
+          if (updated_value != wos->concreteStore[i]) {
```

Comment:
Here is where all the magic happens:

`wos->concreteStore[i]` contains the concrete object-state representation before the call and is copied to the external address before the call. This includes the concrete values representing symbolic expressions.
`update_value` represents the current external value at location `i` after the call. If `update_value` and `wos->concreteStore[i]` differ, we know that the external function has changed the object, so we do need to propagate this back to the object. This is especially important,  if the external function has changed the value that has been previously represented by a symbolic value. In this case we need to update the object state accordingly, therefore we call the `wos->write8()` function. This has been handled incorrectly by KLEE before, i.e. not updating the information that a value is not symbolic anymore, only updating the `concreteStore` representation. This lead to subsequent reads of that object still using the symbolic values instead of the concrete one.

In summary, we have to update the object state's meta-data one way or another.

The major drawback of your solution is that always the full object will be concretised even if only one byte has changed externally.

This PR preserves all the symbolic information of an object that has not been modified by external functions but updates everything else accordingly.
This shouldn't have a higher performance impact for smaller objects.

The performance impact of your suggestion depends on the original state of the object.

Reply:
@MartinNowack thanks for the detailed explanation.  I would add a (shorter) comment along those lines in the text -- the key part would be the first sentence, I think. 

This being said, while your solution is more performant in some cases, it over-approximates things.  That's because it can happen that the external call happened to write the same values for some bytes, and keeping those symbolic would lead to inconsistencies.  It is easy to construct such an example, and might not be so unlikely in practice given that you do this at the byte granularity (and some values, such as 0, might be quite common). 

One solution would be to do this only under the `over-approx` policy I introduced in #1696, what do you think?

Reply:
@ccadar Indeed this is a problem. The only upside is that it's less likely to have an impact as the concretisation (i.e. using the results of the solver call) still has the same result. Therefore, recurring queries with the same constraint set would lead to the same results. 

I like the `over-approx` solution, that would make a cleaner design and we could switch between both implementations easily.

Reply:
Great, then let's guard this by the `--over-approx` flag once #1696 is merged.

Reply:
Sounds like a good plan!